lib/ABI/FunctionType.cpp

```diff
@@ -402,189 +385,404 @@ class ConversionHelper {
     llvm::SmallVector<model::Argument, 8> Result;
 
+    // A flag used to differentiate state of the analysis. It is set to `true`,
+    // if there is at least one used register to the right of the current
+    // one making it very likely to be either used or skipped over as padding.
     bool MustUseTheNextOne = false;
-    auto AllowedRange = llvm::enumerate(llvm::reverse(AllowedRegisters));
-    for (auto Pair : AllowedRange) {
-      size_t Index = AllowedRegisters.size() - Pair.index() - 1;
+
+    // Enumerate all the registers dedicated by the ABI for passing parameters
+    // in reverse order (last register is looked at first).
+    for (auto Pair : llvm::enumerate(llvm::reverse(Allowed))) {
+      size_t Index = Allowed.size() - Pair.index() - 1;
       model::Register::Values Register = Pair.value();
-      bool IsUsed = UsedRegisters.find(Register) != UsedRegisters.end();
-      if (IsUsed) {
-        model::Argument Temporary;
-        if constexpr (!DryRun)
-          Temporary.Type = getTypeOrDefault(UsedRegisters.at(Register).Type,
-                                            Register,
-                                            TheBinary);
-        Temporary.CustomName = UsedRegisters.at(Register).CustomName;
-        Result.emplace_back(Temporary);
+      auto CurrentRegisterIterator = Used.find(Register);
+      if (CurrentRegisterIterator != Used.end()) {
+        // If the current register is confirmed to be in use, convert it into
+        // an argument.
+        auto &Current = Result.emplace_back();
+        Current.Type = ensureTypeIsAvailable(Used.at(Register).Type,
+                                             Register,
+                                             Binary,
+                                             TypeDependencies);
+        Current.CustomName = Used.at(Register).CustomName;
       } else if (MustUseTheNextOne) {
-        if constexpr (!AT::OnlyStartDoubleArgumentsFromAnEvenRegister) {
-          return std::nullopt;
-        } else if ((Index & 1) == 0) {
-          return std::nullopt;
-        } else if (Result.size() > 1 && Index > 1) {
-          auto &First = Result[Result.size() - 1];
-          auto &Second = Result[Result.size() - 2];
-
-          // TODO: see what can be done to preserve names better
-          if (First.CustomName.empty() && !Second.CustomName.empty())
-            First.CustomName = Second.CustomName;
-
-          if constexpr (!DryRun) {
-            auto NewType = buildDoubleType(AllowedRegisters.at(Index - 2),
-                                           AllowedRegisters.at(Index - 1),
-                                           model::PrimitiveTypeKind::Generic,
-                                           TheBinary);
-            if (NewType == std::nullopt)
-              return std::nullopt;
-
-            First.Type = *NewType;
+        // If at least one register to the right of the current one is confirmed
```

Comment:
I'm confused about what's going on in this branch here.
You're trying to merge, say, two 32-bit register arguments into a single 64-bits, right?
Is this required or could we simply emit two 32-bit arguments?

In all this handling for double-register sized arguments, it seems we're not always marking all the register before the last used as arguments under certain conditions.
That is, if `AT::OnlyStartDoubleArgumentsFromAnEvenRegister && Index != 0 && (Index & 1) == 0 && Result.size() > 1 && UpperKind == LowerKind` we do the double register thing, but otherwise?

In x86-64 SystemV, if `rsi` is an argument, then `rdi` is an argument too. In general if `argn` register is an argument all registers before that are arguments too. Is this the case?

Reply:
(I'll start in the opposite order, since it's easier to explain that way).

> In x86-64 SystemV, if `rsi` is an argument, then `rdi` is an argument too. In general if `argn` register is an argument all registers before that are arguments too. Is this the case?

When working from a binary, this piece of logic is already handled on our `Raw` functions within the efa, all the `YesOrDead` and the like states are flagged there using the related part of the pipeline.

> In all this handling for double-register sized arguments, it seems we're not always marking all the register before the last used as arguments under certain conditions.

As such, this part of the conversion is mostly for the hand-written `Raw` function types (since those that went through our usual pipeline can never get caught in it).

Since the ones written manually are probably written this way on purpose, I go to greater length to not discard this information when converting, as in, I see any register skips as intentional, i.e. the padding and make decision based on that.

> You're trying to merge, say, two 32-bit register arguments into a single 64-bits, right?

Yep, under the very specific architectures. Note also, that when splitting this back up in, say, `abi::FunctionType::Layout`, it will be indicated that this "64-bit argument" uses two real registers - hence we don't look the fact that this type is actually composed of two sub-types, half-size each.

Reply:
> When working from a binary, this piece of logic is already handled on our Raw functions within the efa, all the YesOrDead and the like states are flagged there using the related part of the pipeline.

Ah, my brain is melting. Sorry.

> As such, this part of the conversion is mostly for the hand-written Raw function types (since those that went through our usual pipeline can never get caught in it).
>
> Since the ones written manually are probably written this way on purpose, I go to greater length to not discard this information when converting, as in, I see any register skips as intentional, i.e. the padding and make decision based on that.

Hand-written `RawFunctionType`s? The only role in life of `RawFunctionType` is to be generated by `-detect-abi`. Humans will always want to write `CABIFunctionType`s.
Why do you say that `-detect-abi` output will never go through this? If this is the case, it's not very useful.

Reply:
I guess so, but it's also not harmful. Plus it's not like I wrote it now, I just rearranged it somewhat (before there was a `buildDoubleType` function, but I got rid of it because it added the type directly, and since this was the only user, I just ported the body in, nothing more.

Reply:
This is reworked from the ground up. I'm closing this thread as it's no longer relevant.